File: 1_C/Problem Set/credit/credit.c
#include <cs50.h>
#include <stdio.h>

#define THIRTEEN 1000000000000
#define FOURTEEN 10000000000000
#define FIFTEEN 100000000000000
#define SIXTEEN 1000000000000000

string check(long num);
bool AE(long num);
bool MC(long num);
bool Visa(long num);
bool valid(long num);

int main(void)
{
    long num = get_long("Number: ");
    string msg = check(num);
    printf("%s\n", msg);
}

// Determine the last string to print
string check(long num)
{
    if (AE(num) && valid(num))
    {
        return "AMEX";
    }
    else if (MC(num) && valid(num))
    {
        return "MASTERCARD";
    }
    else if (Visa(num) && valid(num))
    {
        return "VISA";
    }
    else
    {
        return "INVALID";
    }
}

// Check if the number likes Amercan Express
bool AE(long num)
{
    if ((num / FIFTEEN < 10) && ((num / FOURTEEN == 34) || (num / FOURTEEN == 37)))
    {
        return true;
    }
    return false;
}

// Check if the number likes MasterCard
bool MC(long num)
{
    if ((num / SIXTEEN == 5) && (num / FIFTEEN % 10 < 6))
    {
        return true;
    }
    return false;
}

// Check if the number likes Visa
bool Visa(long num)
{
    if ((num / THIRTEEN == 4) || (num / SIXTEEN == 4))
    {
        return true;
    }
    return false;
}

// Use Luhn’s Algorithm checks if the number is valid.
bool valid(long num)
{
    long temp = num;
    int sum = 0;
    while (temp)
    {
        int first = temp % 100 / 10 * 2;
        if (first > 9)
        {
            first = first % 10 + first / 10;
        }
        int second = temp % 10;
        sum += first + second;
        temp /= 100;
    }
    return !(sum % 10);
}

File: 6_Python/Problem Set/sentimental-credit/credit.py
# TODO
from cs50 import get_string, get_int

def main():
    number = get_int("Number: ")
    number = str(number)

    if AE(number):
        print('AMEX')
    elif MC(number):
        print('MASTERCARD')
    elif Visa(number):
        print('VISA')
    else:
        print('INVALID')


# Validation of the input number
def validate(number):
    sum = 0
    for i in range(-2, -len(number) - 1, -2):
        tmp = int(number[i]) * 2
        if tmp > 9:
            tmp = int(tmp / 10) + int(tmp % 10)
        sum += tmp
    for i in range(-1, -len(number) - 1, -2):
        sum += int(number[i])
    return not(sum % 10)


# If the input is a American Express number
def AE(number):
    if len(number) != 15 or number[:2] not in ['34', '37']:
        return False
    if not(validate(number)):
        return False
    return True


# If the input is a MasterCard number
def MC(number):
    if len(number) != 16 or number[:2] not in ['51', '52', '53', '54', '55']:
        return False
    if not(validate(number)):
        return False
    return True


# # If the input is a Visa number
def Visa(number):
    if len(number) not in [13, 16] or number[0] != '4':
        return False
    if not(validate(number)):
        return False
    return True


if __name__ == "__main__":
    main()